c10/test/util/bfloat16_test.cpp

```diff
@@ -146,6 +149,54 @@ TEST(BFloat16Math, Subtraction) {
   EXPECT_EQ(res, expected);
 }
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(BFloat16Math, NextAfter) {
```

Comment:
I think this is going to require "golden values" testing because there's no NumPy reference for it. 

One easy way to get started would be with JAX's nextafter operation. I would use it to harvest the values for:

0+ -> 0-
0- -> 0+
0 -> 1
0 -> -1
1 -> 2
1 -> 0
-1 -> -2
-1 -> 0
nan -> 0
0 -> nan
nan -> nan
nan -> inf
inf -> nan
inf -> -inf
-inf -> inf
inf -> 0
0 -> inf
-inf -> 0
0 -> -inf

Plus additional interesting values. In particular where the next value has different exponent bits is super interesting to validate. 

I would do that test in Python and compare for equality with the golden values using atol/rtol=0. How does that sound?

Reply:
Initially I had thought of writing test in Python but generating valid reference `bfloat16` value seemed more painful because I think to represent bfloat16, we will have to use some dummy object which represents `bfloat16` (16-bit entity) while Python double are `double` (so just correctly specifying that value seemed more tricky) (Maybe I am wrong didn't think a lot about it).

It seemed much natural to do so in C++ and hence took this approach. 
Currently samples try to cover all the branches. However adding the samples you mentioned makes sense.

Do we have a strong reason to have this test in Python? (otherwise we may keep this in C++)

Thanks!!

Reply:
I have a general preference for tests in Python for better readability, discoverability, and maintainability.

I don't think you need to worry about a dummy object. When doing "golden value" testing you can just write the exact values as Python literals. For example, let's say 0->1 is .1000768, which it obviously isn't, but then the test might look like:

```
test_cases = (
  (0, 1, .1000768),
)

for a, b, expected in test_cases:
  a = torch.tensor(a, device=device, dtype=torch.bfloat16)
  b = torch.tensor(b, device=device, dtype=torch.bfloat16)
  self.assertEqual(torch.nextafter(a, b).item(), expected)
```

Does that make sense? The dtype of the value you acquire from calling item shouldn't matter, I think

Reply:
Ummm, I can try that 🤔. Thanks!

Reply:
Thanks @mruberry that worked! (we need C++ test for `0` and `-0` because as per IEEE-754 both are equal. So we need a test to check for bitwise-equality with the expected value)

Code (and output) to get the golden values:

<details>

```python
import jax
import jax.numpy as jnp
import numpy as np

dtype=jnp.bfloat16
nan = float('nan')
inf = float('inf')
cases = ((0, -0),
         (-0, 0),
         (0, 1),
         (0, -1),
         (1, -2),
         (1, 0),
         (1, 2),
         (-1, -2),
         (-1, 0),
         (2, -1),
         (2, 1),
         (20, 3000),
         (20, -3000),
         (3000, -20),
        (-3000, 20),
         (nan, 0),
         (0, nan),
         (nan, nan),
         (nan, inf),
         (inf, nan),
         (inf, -inf),
         (-inf, inf),
         (inf, 0),
         (0, inf),
         (-inf, 0),
         (0, -inf),
         )

print("(")
for from_v, to_v in cases:
    from_a = jnp.array([from_v], dtype=dtype)
    to_a = jnp.array([to_v], dtype=dtype)
    result = (from_v, to_v, jnp.nextafter(from_a, to_a).item())
    print(result,",")
print(")")
```

Output
```
(
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
(0, 0, 0.0) ,
(0, 0, 0.0) ,
(0, 1, 9.183549615799121e-41) ,
(0, -1, -9.183549615799121e-41) ,
(1, -2, 0.99609375) ,
(1, 0, 0.99609375) ,
(1, 2, 1.0078125) ,
(-1, -2, -1.0078125) ,
(-1, 0, -0.99609375) ,
(2, -1, 1.9921875) ,
(2, 1, 1.9921875) ,
(20, 3000, 20.125) ,
(20, -3000, 19.875) ,
(3000, -20, 2992.0) ,
(-3000, 20, -2992.0) ,
(nan, 0, nan) ,
(0, nan, nan) ,
(nan, nan, nan) ,
(nan, inf, nan) ,
(inf, nan, nan) ,
(inf, -inf, 3.3895313892515355e+38) ,
(-inf, inf, -3.3895313892515355e+38) ,
(inf, 0, 3.3895313892515355e+38) ,
(0, inf, 9.183549615799121e-41) ,
(-inf, 0, -3.3895313892515355e+38) ,
(0, -inf, -9.183549615799121e-41) ,
)
```

</details>

Reply:
(You may need to set print options to display all the decimals properly to do an exact comparison)

Reply:
Initially, I did try with that but as it turns out, even after setting higher precision JAX doesn't actually display the correct value.

However, Python prints the value correctly (had checked the value `from=0 and to=1` by hand)

Snippet with `precision=16`:

<details>

```python

import jax
import jax.numpy as jnp
import numpy as np

dtype=jnp.bfloat16
nan = float('nan')
inf = float('inf')
cases = ((0, -0),
         (-0, 0),
         (0, 1),
         )

jnp.set_printoptions(precision=16)

print("(")
for from_v, to_v in cases:
    from_a = jnp.array([from_v], dtype=dtype)
    to_a = jnp.array([to_v], dtype=dtype)
    result = (from_v, to_v, jnp.nextafter(from_a, to_a))
    print(result,",")
print(")")
```

Output:
```
(
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
(0, 0, DeviceArray([0], dtype=bfloat16)) ,
(0, 0, DeviceArray([0], dtype=bfloat16)) ,
(0, 1, DeviceArray([9.18355e-41], dtype=bfloat16)) ,
)
```
</details>

Reply:
(failures on windows seemed to be from master)